sql: fix COPY bugs for binary/csv formats

rafiss opened the pull request:
See individual commits:

sql: return correct format code for COPY
fixes #63235 

sql: COPY BINARY messages can be split at arbitrary boundaries 
fixes #68805 

sql: COPY CSV handles arbitrary message boundaries
refs #68805 

sql: fix byte escaping in COPY CSV input 
fixes #68804 

Release justification: Fixes for high-priority bugs in existing functionality. ("Another rule of thumb to consider: if a customer hit this bug after releasing, would we patch the release to fix it? If we would patch the release later to fix the bug, then usually it's best to fix the bug now.")

rafiss committed:
sql: return correct format code for COPY

Fixes #63235

Release note (bug fix): When using COPY FROM .. BINARY, the correct
format code will now be returned.

rafiss committed:
sql: COPY CSV handles arbitrary message boundaries

Release note (bug fix): Previously, COPY FROM ... CSV
would require each CopyData message to be split at the boundary of a
record. This is a bug since the COPY protocol allows messages to be
split at arbitrary points. This is fixed now.

rafiss committed:
sql: COPY BINARY mesages can be split at arbitrary boundaries

Fixes #68805

Release note (bug fix): Previously, COPY FROM ... BINARY would return
an error if the input data was split across different messages. This is
fixed now.

rafiss committed:
sql: fix byte escaping in COPY CSV input

Fixes #68804

Release note (bug fix): COPY FROM ... CSV did not correctly handle
octal byte escape sequences such as `\011` when using a BYTEA column`.
This is now fixed.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/69066)
<!-- Reviewable:end -->

otan commented on pkg/sql/copy.go:
```diff
@@ -387,6 +387,15 @@ func (c *copyMachine) readTextData(ctx context.Context, final bool) (brk bool, e
 }
 
 func (c *copyMachine) readCSVData(ctx context.Context, final bool) (brk bool, err error) {
+	// Don't try to parse the CSV until we read all the data. This is because
```
should we look at improving this in future / make an issue?

rafiss replied:
as @steven-sheehy says this behavior isn't that good, and it might even be better to leave the CSV parsing as-is instead of doing this workaround

i'll give a real fix another shot.

otan commented on pkg/sql/copy.go:
```diff
@@ -424,6 +424,17 @@ func (c *copyMachine) readCSVTuple(ctx context.Context, record []string) error {
 			continue
 		}
+		switch t := c.resultColumns[i].Typ; t.Family() {
+		case types.BytesFamily,
```
where did you get this list from? how do we know it's complete?
(it's weird to me that e.g. Timestamp is ehre but Time is not)

rafiss replied:
the list comes from from what `COPY FROM ... TEXT` is already doing: https://github.com/cockroachdb/cockroach/blob/61fdb08be3835e3c4e18ba6c0389bc7edc43d7d1/pkg/sql/copy.go#L660-L670

it seems old, wouldn't be surprised if it has the wrong types in it

i can extract that into its own function

rafiss replied:
hmmm actually looks like the list we have is already too big -- PG doesn't do this decoding for `TEXT` columns. i think we only want this for `BytesFamily`

rafiss replied:
hmmm actually this is wrong. I tested in PG, and COPY CSV does not do `\xdd` or `\ddd` escaping. Left a comment on https://github.com/cockroachdb/cockroach/issues/68804 explaining why that issue is happening

steven-sheehy commented on pkg/sql/pgwire/testdata/pgtest/copy:
```diff
@@ -310,3 +310,184 @@ ReadyForQuery
+{"Type":"ReadyForQuery","TxStatus":"I"}
+
+# Verify that byte-escape sequences are processed in CSV mode.
+# `MixceDU0` is base64 for `2,\x54`.
```
Does this support octal encoding as well like PostgreSQL [does](https://www.postgresql.org/docs/13/sql-copy.html)? Something like `1,"\157\156\145"`. Would be good to add a test if so.

rafiss replied:
yes it should. adding a test

steven-sheehy commented on pkg/sql/copy.go:
```diff
@@ -383,6 +387,15 @@ func (c *copyMachine) readTextData(ctx context.Context, final bool) (brk bool, e
+	// we don't have a good way of knowing how many bytes to read for the next
+	// record ahead of time (without reimplementing all the CSV parsing logic
+	// here). This means the input buffer will keep growing -- the memory is
+	// accounted in processCopyData.
```
So what will happen if I want to upload a CSV file that's 10GB? It won't process a single row from it until all 10GBs are uploaded to memory? If so, that just trades one problem for another.

rafiss replied:
I agree. I was thinking about how to do this and I think we'll need to fork the csv reader implementation from go. (we'd also need to do that anyway in order to support some of the other COPY options)

one added challenge is that the COPY CSV format requires a newline to appear after each record *except* for the last one. so if we encounter an EOF on an individual CopyData input buffer, it's hard to tell if that's the end of all the CSV input or if we need to check for another input buffer

tomorrow I'll take a look at how postgres does this and explore some other CSV parsers in go.

steven-sheehy replied:
Would it help if I create a separate ticket just for partial records with COPY CSV with the info in my [comment](https://github.com/cockroachdb/cockroach/issues/68805#issuecomment-897891364)? That way this ticket could be merged sooner while the bigger problem with the CSV reader could be worked on separately?

rafiss replied:
yes, sure another ticket would help. if i don't get to this soon then we can go ahead with this PR without the CSV buffering changes

I did find that Postgres solves this by buffering line-by-line, and taking into account whether it is in a quoted field or not: https://github.com/postgres/postgres/blob/9626325da5e8e23ff90091bc96535495d350f06e/src/backend/commands/copyfromparse.c#L1090-L1096

rafiss committed:
sql: add edge-case tests for COPY CSV

These are worth adding since we now have special logic for the CSV
format for finding newlines.

Release note: None

rafiss commented:
This is ready for another look -- I updated the CSV buffering logic so it searches for a non-escaped newline.

otan commented on pkg/sql/copy.go:
```diff
@@ -244,7 +245,9 @@ func (c *copyMachine) run(ctx context.Context) error {
-		c.csvReader = csv.NewReader(&c.buf)
+		c.csvInput.Reset()
+		c.csvReader = csv.NewReader(&c.csvInput)
+		//c.csvReader = csv.NewReader(&c.buf)
```
delete?

rafiss replied:
fixed

otan reviewed (approved):
nice!

rafiss commented:
tftr!

bors r=otan
